File: src/frontend/Symbol.cpp
#include <utility>


#include "Symbol.h"

#include "AST.h"

namespace HDLCompiler {

Symbol::TypeData::TypeData(Symbol::Type type, std::uint64_t width)
	: type(type), width(width) {
	//
}

Symbol::TypeData::TypeData(std::shared_ptr<AST::BlockNode> block, std::uint64_t width)
	: type(Symbol::Type::Block), width(width), block(std::move(block)) {
	//
}

std::string Symbol::TypeData::to_string() const {
	std::string str;

	switch (type) {
		case Type::In:
		    str += "in";
		    break;
		case Type::Out:
		    str += "out";
		    break;
		case Type::Block:
		    str += "block " + block->identifier->value;
		    break;
	}

	if (width != 1) {
		str += "[" + std::to_string(width) + "]";
	}

	return str;
}

Symbol::Symbol(std::string name, std::shared_ptr<Symbol::TypeData> type)
	: name(std::move(name)), type(std::move(type)) {
	//
}

std::ostream& operator<<(std::ostream &stream, const Symbol &symbol) {
	stream << symbol.type->to_string() << " " << symbol.name << " @ " << symbol.signal;
	return stream;
}

}

File: src/frontend/symbol.rs

use std::rc::Weak;
use std::cell::RefCell;

use crate::frontend::ast::{BlockNode, EdgeType};

#[derive(Debug, Clone)]
pub enum SymbolTypeSpecifier {
    Clock(EdgeType),
    In,
    Out,
    Wire,
    Block(Weak<RefCell<BlockNode>>),
}

#[derive(Debug, Clone)]
pub struct SymbolType {
    pub specifier: SymbolTypeSpecifier,
    pub width: u64
}

#[derive(Debug)]
pub struct Symbol {
    pub name: String,
    pub typ: SymbolType,

    /// For most kinds of symbols, `base_signal_id` and `output_base_signal_id` are the same.
    /// For some however, like the outputs of sequential blocks, they are different.
    pub base_signal_id: u32,
    pub output_base_signal_id: u32,
}
